lib/services/directory-synchronizer.js

```diff
@@ -0,0 +1,144 @@
+import { getFiles } from '../util/file';
+import { NTriplesMutator } from './n-triples-mutator';
+import { query } from '../util/database';
+import moment from 'moment';
+import groupBy from 'lodash.groupby';
+import { SemanticFile } from '../entities/semantic-file';
+import { SERVICE_URI } from '../../env';
+import { VersionFile } from '../entities/version-file';
+
+/**
+ * Abstract service providing basic synchronization between files on-disk and in-store
+ */
+export class DirectorySynchronizer {
+
+  constructor(root) {
+    this.debug = false;
+    this.root = root;
+    this.publisher = `${SERVICE_URI}${root}`;
+    this.files = [];
+    this.mutator = new NTriplesMutator();
+  }
+
+  /**
+   * Synchronize the files saved in-store with the once saved on-disk.
+   *
+   * @returns {Promise<{missing, finished}>}
+   */
+  async sync() {
+    const onDisk = await this.getSemanticFilesOnDisk();
+    const inStore = await this.getSemanticFilesInStore();
+    let {finished, missing} = await this.doSync(onDisk, inStore);
+
+    // NOTE: we group the found files locally by creation date
+    this.files = finished;
+
+    // NOTE: gen. small console report on what occurred
+    if (this.debug) {
+      const rf = finished.map(file => new Object({status: 'DONE', file}));
+      const rm = missing.map(file => new Object({status: 'LOST', file}));
```

Comment:
What do `rm` and `rf` stand for ?

Reply:
me being lazy, I think they stand for `reportFinished` and `reportMissing` as it was just to be used in logging.